tests/test-umockdev-vala.vala

```diff
@@ -855,6 +855,86 @@ t_mt_uevent ()
   assert_cmpuint (change_count, CompareOperator.EQ, num_changes);
 }
 
+static bool
+ioctl_custom_handle_ioctl_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    if (client.request == 1) {
+      client.complete(*(long*)client.arg.data, 0);
+    } else if (client.request == 2) {
+      client.complete(-1, Posix.ENOMEM);
+    } else {
+      client.complete(-5, 1);
+    }
+    return true;
+}
+
+static bool
+ioctl_custom_handle_write_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    client.set_data("written", client.arg);
+    client.complete(client.arg.data.length, 0);
+    return true;
+}
+
+static bool
+ioctl_custom_handle_read_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    UMockdev.IoctlData written = client.get_data("written");
+
+    assert(written != null);
```

Comment:
Any real code should subclass `IoctlBase` and keep its data around that way.

The main issue would be keeping data per-connection (i.e. `IoctlClient`). We don't have a convenient way to do so other than using `g_object_set_data` right now or keeping around a hash table which really isn't better.

I can add some sanity checks and comments just in case someone gets the idea of copying this code. Ideally they would be copying the SPI or pcap code though.